### ISSUE TITLE
Training for two rounds is not equivalent to training one round twice

### ISSUE THREAD

**Issue by kobics** [microsoft/LightGBM#3247]
## How you are using LightGBM?

LightGBM component: 2.3.1

## Environment info

Operating System:

CPU/GPU model: CPU

C++ compiler version: NA

CMake version: NA

Java version: NA

Python version: 3.7.6

R version: NA

Other: NA

LightGBM version or commit hash:

2.3.1

## Steps to reproduce

1.  build a model with , num_boost_round=1, store the model in a variable called model_a1
2. build a model with , num_boost_round=1, init_model=model1.  store the model in a variable called model_a2
3. build a model with, num_boost_round=2, store the model in a variable called model_b2
4. predictions of model_a2 and model_b2 are not the same

**Comment by PastelBelem8** [comment]
Have you tried setting the seed?

**Comment by guolinke** [comment]
@kobics there are many factors will cause this. 
please provide a reproducibly example, by random gerenated data.

**Comment by kobics** [comment]
> @kobics there are many factors will cause this.
> please provide a reproducibly example, by random gerenated data.

import numpy as np
import lightgbm as lgb

X = np.random.randn(100,3)
Y = np.random.randn(100)
S = lgb.Dataset(X,Y,free_raw_data=False)

params = {
'task':'train',
'boosting_type':'gdbt',
'objective':'regression_l2',
'metric':{'l2'},
'max_depth':2,
'learning_rate':0.1}

# 1 round
model1 = lgb.train(params,S,num_boost_round=1,valid_sets=[S],early_stopping_rounds=10)
model1.save_model('model1.txt')

# 1 more round
model2 = lgb.train(params,S,num_boost_round=1,valid_sets=[S],early_stopping_rounds=10, init_model=model1)
model3 = lgb.train(params,S,num_boost_round=1,valid_sets=[S],early_stopping_rounds=10, init_model='model1.txt')

# two rounds
model4 = lgb.train(params,S,num_boost_round=2,valid_sets=[S],early_stopping_rounds=10)

I was expecting model2, model3, model4 to be the same. model2 and model3 are taking one more round over model1, and model4 is generated with 2 rounds. I tried to set a seed, and was not able to make it work.

**Comment by guolinke** [comment]
can you try
```
model1 = lgb.train(params,S,num_boost_round=1,valid_sets=[S],early_stopping_rounds=10)
model1.save_model('model1.txt')
S = lgb.Dataset(X,Y,free_raw_data=False)
model2 = lgb.train(params,S,num_boost_round=1,valid_sets=[S],early_stopping_rounds=10, init_model=model1)
S = lgb.Dataset(X,Y,free_raw_data=False)
model3 = lgb.train(params,S,num_boost_round=1,valid_sets=[S],early_stopping_rounds=10, init_model='model1.txt')
S = lgb.Dataset(X,Y,free_raw_data=False)
model4 = lgb.train(params,S,num_boost_round=2,valid_sets=[S],early_stopping_rounds=10)
```
?

**Comment by kobics** [comment]
> can you try
> 
> ```
> model1 = lgb.train(params,S,num_boost_round=1,valid_sets=[S],early_stopping_rounds=10)
> model1.save_model('model1.txt')
> S = lgb.Dataset(X,Y,free_raw_data=False)
> model2 = lgb.train(params,S,num_boost_round=1,valid_sets=[S],early_stopping_rounds=10, init_model=model1)
> S = lgb.Dataset(X,Y,free_raw_data=False)
> model3 = lgb.train(params,S,num_boost_round=1,valid_sets=[S],early_stopping_rounds=10, init_model='model1.txt')
> S = lgb.Dataset(X,Y,free_raw_data=False)
> model4 = lgb.train(params,S,num_boost_round=2,valid_sets=[S],early_stopping_rounds=10)
> ```
> 
> ?

works! thanks.
can you explain why?

**Comment by guolinke** [comment]
Dataset will record the `init_score`, which is from `init_model`. 
So in continued train mode, it is better to reset Dataset.
You can also clear the `init_score` in Dataset as well, by setting its values to zeros.
